### ISSUE TITLE
Jump and Ramp_fit steps don't handle subarray readnoise and gain ref files

### ISSUE THREAD

**Issue by hbushouse** [STScI-JWST/jwst#1108]
NIRSpec has begun delivering subarray-specific reference files for both readnoise and gain. The jump and ramp_fit steps are currently setup in a way that assumes those reference data arrays will always be full-frame and hence take a very simple approach to extracting a subarray from those reference arrays to match a science exposure subarray. These need to be updated to handle reference files that are themselves subarrays, in the way some of the other steps like dq_init and saturation handle subarray reference files.

**Comment by hbushouse** [comment]
A typical error resulting from this problem is:

ERROR RUNNING STEP 'SloperPipeline':
    operands could not be broadcast together with shapes
    (2,12,256,2048) (0,2048) (2,12,256,2048)
Traceback (most recent call last):
  File "/Users/bushouse/anaconda3/envs/myjwst/lib/python2.7/site-packages/jwst-0.7.8rc1.dev307-py2.7-macosx-10.7-x86_64.egg/jwst/stpipe/cmdline.py", line 320, in step_from_cmdline
    step.run(*positional)
  File "/Users/bushouse/anaconda3/envs/myjwst/lib/python2.7/site-packages/jwst-0.7.8rc1.dev307-py2.7-macosx-10.7-x86_64.egg/jwst/stpipe/step.py", line 361, in run
    result = self.process(*args)
  File "/Users/bushouse/anaconda3/envs/myjwst/lib/python2.7/site-packages/jwst-0.7.8rc1.dev307-py2.7-macosx-10.7-x86_64.egg/jwst/pipeline/calwebb_sloper.py", line 106, in process
    input = self.jump(input)
  File "/Users/bushouse/anaconda3/envs/myjwst/lib/python2.7/site-packages/jwst-0.7.8rc1.dev307-py2.7-macosx-10.7-x86_64.egg/jwst/stpipe/step.py", line 361, in run
    result = self.process(*args)
  File "/Users/bushouse/anaconda3/envs/myjwst/lib/python2.7/site-packages/jwst-0.7.8rc1.dev307-py2.7-macosx-10.7-x86_64.egg/jwst/jump/jump_step.py", line 69, in process
    rej_thresh, do_yint, sig_thresh)
  File "/Users/bushouse/anaconda3/envs/myjwst/lib/python2.7/site-packages/jwst-0.7.8rc1.dev307-py2.7-macosx-10.7-x86_64.egg/jwst/jump/jump.py", line 81, in detect_jumps
    data *= gain_2d
ValueError: operands could not be broadcast together with shapes (2,12,256,2048) (0,2048) (2,12,256,2048)

**Comment by hbushouse** [comment]
Actually, upon closer examination of other steps like dq_init and saturation, it appears that all of these may have the same problem. They all only look at the SUBSTRTn values in the science data and then extract that region from the reference data array, which only works properly if the reference data array is full-frame. These all need to be updated to also consider the SUBSTRTn values of the **_reference data_**, when figuring out the slice to extract.

**Comment by jdavies-st** [comment]
Could there be a generalized function available to all pipeline steps that returns the part of the reffile subarray that matches an input science image?  So that we wouldn't have to write the same thing for each step that uses a subarray from a reffile, or a subarray from a subarray reffile.  Something like:
```
def cut_out_reffile_subarray(reference_image, science_image):
    """Cut out the part of the reffile data array that corresponds to the input science data array

    Parameters
    -------------
    reference_image: datamodel
        reference datamodel from which the cutout will be made

    science_image : datamodel
        science datamodel that the reffile is being matched to

    Returns
    ---------
    np.ndarray object
    """
    pass
```
Or is the slice or subarray needed from a given reference file too step- or reffile type-specific?

**Comment by philhodge** [comment]
I do think this is practical, and I think it's what we should do.

One possibility would be to write a function that returns a tuple of slice objects, which could then be applied to a reference file to get the matching region.

**Comment by hbushouse** [comment]
Yes, we definitely need to (finally!) have a common library routine to do this once and for all from all cal steps that need it. We've got a lib directory now, let's use it.

**Comment by hbushouse** [comment]
Fixed by #1106 and #1156.
